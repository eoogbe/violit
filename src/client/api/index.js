/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
