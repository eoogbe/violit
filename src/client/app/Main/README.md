```jsx
import { MemoryRouter } from 'react-router-dom';
import * as BOARD from '__fixtures__/boardFixture';
import * as THREADS from '__fixtures__/threadsFixture';

mockFetch([
  BOARD.boardFetch,
  THREADS.threadsFetch,
]);

<StoreProvider>
  <MemoryRouter>
    <Main />
  </MemoryRouter>
</StoreProvider>
```

When error:

```jsx
import { MemoryRouter } from 'react-router-dom';
import * as BOARD from '__fixtures__/boardFixture';
import * as THREADS from '__fixtures__/threadsFixture';

mockFetch([
  BOARD.boardFetch,
  THREADS.threadsFetch,
]);

const state = {
  error: true,
};

<StoreProvider state={state}>
  <MemoryRouter>
    <Main />
  </MemoryRouter>
</StoreProvider>
```
