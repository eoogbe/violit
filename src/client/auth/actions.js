/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import createActions from 'client/utils/createActions';

export default createActions('auth', [
  'LOAD_TRIGGERED',
  ['LOAD_SUCCEEDED', 'user'],
  'LOAD_FAILED',
  ['LOGIN_TRIGGERED', 'credentials'],
  ['LOGIN_SUCCEEDED', 'user'],
  'LOGIN_FAILED',
  ['CREDENTIALS_INVALID', 'error'],
  'LOGOUT_TRIGGERED',
  ['SIGNUP_TRIGGERED', 'user'],
  ['SIGNUP_SUCCEEDED', 'user'],
  ['SIGNUP_FAILED'],
  ['CHECK_HAS_USER_TRIGGERED', 'username'],
  ['CHECK_HAS_USER_SUCCEEDED', 'hasUser'],
  'CHECK_HAS_USER_FAILED',
]);
