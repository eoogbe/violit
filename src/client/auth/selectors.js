/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import { createSelector } from 'reselect';
import { NAME } from './constants';

/** Selects the auth state. */
export const getAuth = R.prop(NAME);

/** Selects the credentials of the current user. */
export const getCredentials = createSelector(
  [getAuth],
  R.prop('credentials')
);

/** Selects the username of the current user. */
export const getUsername = createSelector(
  [getAuth],
  R.path(['credentials', 'username'])
);
