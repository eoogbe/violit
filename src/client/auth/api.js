/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import { checkExists, href, load, sendCreate, sendDelete } from 'client/api';

/** Creates the specified `credentials` on the server. */
const createCredentials = async (url, credentials, {
  status,
  errorMessage,
 }) => {
  try {
    const { body } = await sendCreate(
      url,
      ['username', 'password'],
      credentials
    );
    return R.pick(['username'], body);
  } catch (err) {
    if (R.path(['response', 'status'], err) !== status) {
      throw new Error(errorMessage);
    }

    const message = R.path(['response', 'body', 'error', 'message'], err);
    if (!message) {
      throw new Error(errorMessage);
    }

    return { error: message };
  }
};

/** Loads the current user's credentials from the server if they exist. */
export const loadCredentials = async () => {
  try {
    const credentials = await load(href('/auth'));
    return R.pick(['username'], credentials);
  } catch (err) {
    if (R.path(['response', 'status'], err) !== 405) {
      throw new Error('Something went wrong with loading the credentials.');
    }

    return null;
  }
};

/** Authenticates a user. */
export const login = (credentials) =>
  createCredentials(href('/auth'), credentials, {
    status: 404,
    errorMessage: 'Something went wrong with logging in.'
  });

/** Logs a user out of the current session. */
export const logout = async () => {
  await sendDelete(href('/auth'));
  return null;
};

/** Registers the specified `user` on the server. */
export const signup = (user) =>
  createCredentials(href('/users'), user, {
    status: 422,
    errorMessage: 'Something went wrong with signing up.'
  });

/** Checks if a user with the specified `username` exists on the server. */
export const checkHasUser = (username) =>
  checkExists(href(`/users/${username}`));
