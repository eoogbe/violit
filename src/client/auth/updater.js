/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  concat,
  constantState,
  decorate,
  handleAction,
  updateStateAt,
  withDefaultState,
} from 'redux-fp';
import matchActions from 'client/utils/matchActions';
import actions from './actions';

const INITIAL_STATE = {};

/** Updater that sets the current user's credentials. */
const setCredentials =
  decorate(
    matchActions([
      actions.loadSucceeded.type,
      actions.loginSucceeded.type,
      actions.signupSucceeded.type,
    ]),
    updateStateAt('credentials'),
    ({ payload }) => () => payload.user
  );

/** Updater that unsets the current user's credentials. */
const unsetCredentials =
  decorate(
    handleAction(actions.logoutTriggered.type),
    updateStateAt('credentials'),
    constantState(null)
  );

/** The auth updater. */
export default decorate(
  withDefaultState(INITIAL_STATE),
  concat(
    setCredentials,
    unsetCredentials,
  )
);
