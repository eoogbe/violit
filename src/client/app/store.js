/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createPromiseListener from 'redux-promise-listener';
import createSagaMiddleware from 'redux-saga';
import saga from './saga';
import updater from './updater';

export const promiseListener = createPromiseListener();

/** Configures the Redux store. */
export default () => {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    (state, action) => updater(action)(state),
    composeWithDevTools(
      applyMiddleware(sagaMiddleware, promiseListener.middleware)
    )
  );
  sagaMiddleware.run(saga);
  return store;
};
