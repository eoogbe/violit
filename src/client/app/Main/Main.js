/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { useSelector } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import styled from 'styled-components';
import { Login, Signup, UnauthenticatedRoute } from 'client/auth';
import { Board } from 'client/board';
import { Alert } from 'client/core';
import { Thread, ThreadForm } from 'client/threads';
import { promiseListener } from '../store';
import { getError } from '../selectors';

const StyledMain = styled.main`
  font-size: 1rem;
  flex: 1;
  position: relative;
  margin: ${({ theme }) => theme.spacingLg};
  background-color: ${({ theme }) => theme.white};
`;

/** The main section of a page. */
const Main = () => {
  const error = useSelector(getError);

  return (
    <StyledMain>
      <Switch>
        <Route exact path="/">
          <Board />
        </Route>
        <Route path="/threads/:slug">
          <Thread />
        </Route>
        <Route path="/new-thread">
          <ThreadForm />
        </Route>
        <UnauthenticatedRoute path="/login">
          <Login createAsyncFunction={promiseListener.createAsyncFunction} />
        </UnauthenticatedRoute>
        <UnauthenticatedRoute path="/signup">
          <Signup createAsyncFunction={promiseListener.createAsyncFunction} />
        </UnauthenticatedRoute>
      </Switch>
      {error && <Alert />}
    </StyledMain>
  );
};

export default Main;
