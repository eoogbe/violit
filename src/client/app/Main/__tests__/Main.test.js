/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import * as BOARD from '__fixtures__/boardFixture';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';
import { constants as authConstants } from 'client/auth';
import mockFetch from 'testUtils/mockFetch';
import { wait } from 'testUtils/wait';
import { promiseListener } from '../../store';
import Main from '../Main';

const mockStore = configureStore([promiseListener.middleware]);

const createMocks = (initialState) => {
  Date.now = jest.fn().mockReturnValue(1572566400000);
  mockFetch([
    BOARD.boardFetch,
    THREADS.threadsFetch,
    THREADS.threadFetch1,
    THREADS.canCreateThreadFetch,
    COMMENTS.canCreateCommentFetch,
    COMMENTS.commentsFetch,
  ]);
  return mockStore(initialState);
};

const expectState = async (initialState, initialEntries) => {
  const store = createMocks(initialState);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={initialEntries}>
          <Main />
        </MemoryRouter>
      </Provider>
    );

    await wait();
  });

  return expect(renderer.toJSON()).toMatchSnapshot();
};

it('renders without crashing', () => {
  return expectState();
});

it('renders the error state', () => {
  return expectState({
    error: true,
  });
});

it('renders the thread screen', () => {
  return expectState({
    [authConstants.NAME]: {
      credentials: USER.credentials,
    },
  }, [THREADS.thread1.path]);
});

it('renders the new thread screen', () => {
  return expectState({}, ['/new-thread']);
});

it('renders the login screen', () => {
  return expectState({}, ['/login']);
});

it('renders the signup screen', () => {
  return expectState({}, ['/signup']);
});

it('with credentials redirects login screen to the home screen', async () => {
  const store = createMocks({
    [authConstants.NAME]: {
      credentials: USER.credentials,
    },
  });

  let location;
  await TestRenderer.act(async () => {
    TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/login']}>
          <Main />
          <Route
            path="*"
            render={(props) => {
              location = props.location;
              return null;
            }}
          />
        </MemoryRouter>
      </Provider>
    );
  });

  expect(location).toHaveProperty('pathname', '/');
});

it('with credentials redirects signup screen to the home screen', async () => {
  const store = createMocks({
    [authConstants.NAME]: {
      credentials: USER.credentials,
    },
  });

  let location;
  await TestRenderer.act(async () => {
    TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/signup']}>
          <Main />
          <Route
            path="*"
            render={(props) => {
              location = props.location;
              return null;
            }}
          />
        </MemoryRouter>
      </Provider>
    );
  });

  expect(location).toHaveProperty('pathname', '/');
});
