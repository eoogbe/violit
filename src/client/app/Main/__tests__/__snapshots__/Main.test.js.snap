// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the error state 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <article
    className="sc-fBuWsC bJYtKI"
  >
    <header
      className="sc-fMiknA kcrmtG"
    >
      <h2
        className="sc-dVhcbM hquhcZ"
      >
        dance
      </h2>
      <a
        href="/new-thread"
        onClick={[Function]}
      >
        <svg
          color="#7F00FF"
          fill="currentColor"
          height="1.5em"
          size="1.5em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "#7F00FF",
            }
          }
          title="New thread"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </a>
    </header>
    <ul
      className="sc-jhAzac gBCyfg"
    >
      <li
        className="sc-Rmtcm hQsHdC"
      >
        <a
          className="sc-bRBYWo hGkBnq"
          href="/threads/lorem-ipsum"
          onClick={[Function]}
        >
          <article
            itemScope={true}
            itemType="http://schema.org/DiscussionForumPosting"
          >
            <header
              className="sc-jKJlTe dCQNcG"
            >
              <span
                className="sc-eNQAEJ jsbjCq"
                itemProp="author"
              >
                fred-astaire
              </span>
              <time
                dateTime="2019-09-29T01:23:45.679Z"
                itemProp="dateCreated"
                title="2019-09-29T01:23:45.679Z"
              >
                1 month ago
              </time>
            </header>
            <p
              className="sc-hzDkRC knCefW"
              itemProp="headline"
            >
              Lorem ipsum
            </p>
          </article>
        </a>
      </li>
      <li
        className="sc-Rmtcm hQsHdC"
      >
        <a
          className="sc-bRBYWo hGkBnq"
          href="/threads/commodo-consequat"
          onClick={[Function]}
        >
          <article
            itemScope={true}
            itemType="http://schema.org/DiscussionForumPosting"
          >
            <header
              className="sc-jKJlTe dCQNcG"
            >
              <span
                className="sc-eNQAEJ jsbjCq"
                itemProp="author"
              >
                fred-astaire
              </span>
              <time
                dateTime="2019-09-27T03:21:09.876Z"
                itemProp="dateCreated"
                title="2019-09-27T03:21:09.876Z"
              >
                1 month ago
              </time>
            </header>
            <p
              className="sc-hzDkRC knCefW"
              itemProp="headline"
            >
              Commodo consequat
            </p>
          </article>
        </a>
      </li>
    </ul>
  </article>
  <div
    aria-live="assertive"
    className="sc-bdVaJa dXBAhf"
    role="alert"
  >
    <p
      className="sc-htpNat bhpHJD"
    >
      Oops! Something went wrong with 
      that action
      !Â 
      <button
        className="sc-bxivhb ErYiX"
        onClick={[Function]}
        type="button"
      >
        Reload page
      </button>
    </p>
    <button
      className="sc-bwzfXH haUZkm"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="1.25em"
        size="1.25em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        title="Dismiss alert"
        viewBox="0 0 24 24"
        width="1.25em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</main>
`;

exports[`renders the login screen 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <form
    className="sc-dnqmqq bwQyKk"
    onSubmit={[Function]}
  >
    <h2
      className="sc-iwsKbI eMpqKh"
    >
      Login
    </h2>
    <ul
      className="sc-gZMcBi dOixHG"
    >
      <li>
        <label
          htmlFor="username"
        >
          Username
        </label>
        <input
          aria-required="true"
          autoFocus={true}
          className="sc-htoDjs kEmxlR"
          id="username"
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </li>
      <li>
        <label
          htmlFor="password"
        >
          Password
        </label>
        <input
          aria-required="true"
          className="sc-htoDjs kEmxlR"
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </li>
    </ul>
    <button
      className="sc-ifAKCX hbBAat sc-gqjmRU kZCXhP"
      disabled={true}
      type="submit"
    >
      Log in
    </button>
  </form>
</main>
`;

exports[`renders the new thread screen 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <form
    className="sc-iRbamj dfRSUU"
    onSubmit={[Function]}
  >
    <h2
      className="sc-jlyJG htSWWt"
    >
      New thread
    </h2>
    <ul
      className="sc-gipzik eXbhrh"
    >
      <li>
        <label
          htmlFor="headline"
        >
          Title
        </label>
        <input
          aria-required="true"
          autoFocus={true}
          className="sc-htoDjs kEmxlR"
          id="headline"
          name="headline"
          onChange={[Function]}
          type="text"
          value=""
        />
      </li>
      <li>
        <label
          htmlFor="article-body"
        >
          Body
        </label>
        <textarea
          className="sc-htoDjs kEmxlR"
          id="article-body"
          name="article-body"
          onChange={[Function]}
          rows={5}
          value=""
        />
      </li>
    </ul>
    <button
      className="sc-ifAKCX hbBAat sc-csuQGl cDFWor"
      disabled={true}
      type="submit"
    >
      Post
    </button>
  </form>
</main>
`;

exports[`renders the signup screen 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <form
    className="sc-VigVT hehCXF"
    onSubmit={[Function]}
  >
    <h2
      className="sc-jTzLTM bADGtk"
    >
      Signup
    </h2>
    <ul
      className="sc-fjdhpX dGhdaN"
    >
      <li>
        <label
          htmlFor="username"
        >
          Username
        </label>
        <input
          aria-required="true"
          autoFocus={true}
          className="sc-htoDjs kEmxlR"
          id="username"
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </li>
      <li>
        <label
          htmlFor="password"
        >
          Password
        </label>
        <input
          aria-required="true"
          className="sc-htoDjs kEmxlR"
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </li>
      <li>
        <label
          htmlFor="password-confirmation"
        >
          Password confirmation
        </label>
        <input
          aria-required="true"
          className="sc-htoDjs kEmxlR"
          id="password-confirmation"
          name="password-confirmation"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </li>
    </ul>
    <button
      className="sc-ifAKCX hbBAat sc-jzJRlG grlris"
      disabled={true}
      type="submit"
    >
      Sign up
    </button>
  </form>
</main>
`;

exports[`renders the thread screen 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <article
    className="sc-jDwBTQ gWsmob"
    itemScope={true}
    itemType="http://schema.org/DiscussionForumPosting"
  >
    <div
      className="sc-gPEVay dOeLqP"
    >
      <h2
        className="sc-brqgnP kAzWxO"
        itemProp="headline"
      >
        Lorem ipsum
      </h2>
      <header
        className="sc-jKJlTe dCQNcG"
      >
        <span
          className="sc-eNQAEJ jsbjCq"
          itemProp="author"
        >
          fred-astaire
        </span>
        <time
          dateTime="2019-09-29T01:23:45.679Z"
          itemProp="dateCreated"
          title="2019-09-29T01:23:45.679Z"
        >
          1 month ago
        </time>
      </header>
      <div
        itemProp="articleBody"
      >
        <div
          className="sc-bZQynM kESxfG"
        >
          <p>
            Dolor sit amet, consectetur adipiscing elit.
          </p>
        </div>
      </div>
      <form
        className="sc-cSHVUG jxHEP"
        onSubmit={[Function]}
      >
        <ul
          className="sc-kAzzGY ckrMOY"
        >
          <li>
            <textarea
              className="sc-htoDjs kEmxlR"
              name="text"
              onChange={[Function]}
              placeholder="Write a comment"
              rows={5}
              value=""
            />
          </li>
        </ul>
        <button
          className="sc-ifAKCX hbBAat sc-chPdSV bBcFRQ"
          disabled={true}
          type="submit"
        >
          Comment
        </button>
      </form>
      <footer
        className="sc-kpOJdX fzEkOA"
      >
        <nav
          className="sc-dxgOiQ qAjpS"
        >
          <button
            aria-controls="overflow-menu-thread-thread1"
            aria-expanded={false}
            aria-haspopup="true"
            className="sc-ckVGcZ ehJtkF"
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="currentColor"
              focusable={false}
              height="1.5em"
              size="1.5em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              title="More actions"
              viewBox="0 0 24 24"
              width="1.5em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </button>
        </nav>
      </footer>
    </div>
    <ul
      className="sc-cvbbAY HtGJA"
    >
      <li
        className="sc-iAyFgw eoNjzy"
        itemProp="comment"
        itemScope={true}
        itemType="http://schema.org/Comment"
      >
        <article>
          <header
            className="sc-jKJlTe dCQNcG"
          >
            <span
              className="sc-eNQAEJ jsbjCq"
              itemProp="author"
            >
              fred-astaire
            </span>
            <time
              dateTime="2019-10-09T10:11:12.123Z"
              itemProp="dateCreated"
              title="2019-10-09T10:11:12.123Z"
            >
              3 weeks ago
            </time>
          </header>
          <div
            itemProp="text"
          >
            <div
              className="sc-bZQynM kESxfG"
            >
              <p>
                Sed do eiusmod tempor incididunt ut labore et dolore.
              </p>
            </div>
          </div>
          <footer
            className="sc-kpOJdX fzEkOA"
          >
            <nav
              className="sc-dxgOiQ qAjpS"
            >
              <button
                aria-controls="overflow-menu-comment-comment1"
                aria-expanded={false}
                aria-haspopup="true"
                className="sc-ckVGcZ ehJtkF"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  focusable={false}
                  height="1.5em"
                  size="1.5em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  title="More actions"
                  viewBox="0 0 24 24"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </button>
            </nav>
          </footer>
        </article>
      </li>
      <li
        className="sc-iAyFgw eoNjzy"
        itemProp="comment"
        itemScope={true}
        itemType="http://schema.org/Comment"
      >
        <article>
          <header
            className="sc-jKJlTe dCQNcG"
          >
            <span
              className="sc-eNQAEJ jsbjCq"
              itemProp="author"
            >
              fred-astaire
            </span>
            <time
              dateTime="2019-10-08T07:08:09.101Z"
              itemProp="dateCreated"
              title="2019-10-08T07:08:09.101Z"
            >
              3 weeks ago
            </time>
          </header>
          <div
            itemProp="text"
          >
            <p
              className="sc-hSdWYo kGFdpB"
            >
              [deleted]
            </p>
          </div>
        </article>
      </li>
      <li
        className="sc-iAyFgw eoNjzy"
        itemProp="comment"
        itemScope={true}
        itemType="http://schema.org/Comment"
      >
        <article>
          <header
            className="sc-jKJlTe dCQNcG"
          >
            <span
              className="sc-eNQAEJ jsbjCq"
              itemProp="author"
            >
              fred-astaire
            </span>
            <time
              dateTime="2019-09-30T01:02:03.456Z"
              itemProp="dateCreated"
              title="2019-09-30T01:02:03.456Z"
            >
              1 month ago
            </time>
          </header>
          <div
            itemProp="text"
          >
            <div
              className="sc-bZQynM kESxfG"
            >
              <p>
                Quis nostrud exercitation ullamco laboris nisi ut.
              </p>
            </div>
          </div>
          <footer
            className="sc-kpOJdX fzEkOA"
          >
            <nav
              className="sc-dxgOiQ qAjpS"
            >
              <button
                aria-controls="overflow-menu-comment-comment3"
                aria-expanded={false}
                aria-haspopup="true"
                className="sc-ckVGcZ ehJtkF"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  focusable={false}
                  height="1.5em"
                  size="1.5em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  title="More actions"
                  viewBox="0 0 24 24"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </button>
            </nav>
          </footer>
        </article>
      </li>
    </ul>
  </article>
</main>
`;

exports[`renders without crashing 1`] = `
<main
  className="sc-eqIVtm kRffCT"
>
  <article
    className="sc-fBuWsC bJYtKI"
  >
    <header
      className="sc-fMiknA kcrmtG"
    >
      <h2
        className="sc-dVhcbM hquhcZ"
      >
        dance
      </h2>
      <a
        href="/new-thread"
        onClick={[Function]}
      >
        <svg
          color="#7F00FF"
          fill="currentColor"
          height="1.5em"
          size="1.5em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "#7F00FF",
            }
          }
          title="New thread"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </a>
    </header>
    <ul
      className="sc-jhAzac gBCyfg"
    >
      <li
        className="sc-Rmtcm hQsHdC"
      >
        <a
          className="sc-bRBYWo hGkBnq"
          href="/threads/lorem-ipsum"
          onClick={[Function]}
        >
          <article
            itemScope={true}
            itemType="http://schema.org/DiscussionForumPosting"
          >
            <header
              className="sc-jKJlTe dCQNcG"
            >
              <span
                className="sc-eNQAEJ jsbjCq"
                itemProp="author"
              >
                fred-astaire
              </span>
              <time
                dateTime="2019-09-29T01:23:45.679Z"
                itemProp="dateCreated"
                title="2019-09-29T01:23:45.679Z"
              >
                1 month ago
              </time>
            </header>
            <p
              className="sc-hzDkRC knCefW"
              itemProp="headline"
            >
              Lorem ipsum
            </p>
          </article>
        </a>
      </li>
      <li
        className="sc-Rmtcm hQsHdC"
      >
        <a
          className="sc-bRBYWo hGkBnq"
          href="/threads/commodo-consequat"
          onClick={[Function]}
        >
          <article
            itemScope={true}
            itemType="http://schema.org/DiscussionForumPosting"
          >
            <header
              className="sc-jKJlTe dCQNcG"
            >
              <span
                className="sc-eNQAEJ jsbjCq"
                itemProp="author"
              >
                fred-astaire
              </span>
              <time
                dateTime="2019-09-27T03:21:09.876Z"
                itemProp="dateCreated"
                title="2019-09-27T03:21:09.876Z"
              >
                1 month ago
              </time>
            </header>
            <p
              className="sc-hzDkRC knCefW"
              itemProp="headline"
            >
              Commodo consequat
            </p>
          </article>
        </a>
      </li>
    </ul>
  </article>
</main>
`;
